framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java

```diff
@@ -171,15 +182,41 @@ public Void visitDeclared(AnnotatedDeclaredType type, Void p) {
          * If any constant-value annotation has &gt; MAX_VALUES number of values provided, treats
          * the value as UnknownVal. Works together with ValueVisitor.visitAnnotation, which issues a
          * warning to the user in this case.
+         *
+         * <p>Additional rules applied for IntVal and IntRange annotations: if anno is IntVal and
+         * the number of values provided is &gt; MAX_VALUES, replace with IntRange; if anno is
+         * IntRange and the range is &le; MAX_VALUES, replace with IntVal
          */
-        private void replaceWithUnknownValIfTooManyValues(AnnotatedTypeMirror atm) {
+        private void replaceWithNewAnnoInSpecialCases(AnnotatedTypeMirror atm) {
             AnnotationMirror anno = atm.getAnnotationInHierarchy(UNKNOWNVAL);
 
             if (anno != null && anno.getElementValues().size() > 0) {
-                List<Object> values =
-                        AnnotationUtils.getElementValueArray(anno, "value", Object.class, false);
-                if (values != null && values.size() > MAX_VALUES) {
-                    atm.replaceAnnotation(UNKNOWNVAL);
+                if (AnnotationUtils.areSameByClass(anno, IntRange.class)) {
+                    Range range = getIntRange(anno);
+                    if (range.to > range.from && !range.isWiderThan(MAX_VALUES)) {
+                        List<Long> values = new ArrayList<>();
+                        for (long value = range.from; value <= range.to; value++) {
+                            values.add(value);
+                        }
+                        atm.replaceAnnotation(createIntValAnnotation(values));
+                    }
+                } else if (AnnotationUtils.areSameByClass(anno, IntVal.class)) {
+                    List<Long> values =
+                            AnnotationUtils.getElementValueArray(anno, "value", Long.class, true);
+                    if (values != null && values.size() > MAX_VALUES) {
+                        long annoMinVal = Collections.min(new ArrayList<>(values));
+                        long annoMaxVal = Collections.max(new ArrayList<>(values));
+                        atm.replaceAnnotation(
+                                createIntRangeAnnotation(new Range(annoMinVal, annoMaxVal)));
+                    }
+
+                } else {
+                    List<Object> values =
```

Comment:
It could be @DoubleVal, @BoolVal, @StringVal or @ArrayLen, so needs to be taken care of.